feat(docs): autogenerating cli references spike

sklppy88 opened the pull request:
This defines both the template, and the method.

The template is simple, and uses one page.

At the entrypoint we have `aztec`

#command (aztec)
## table of contents
## command detail 1
### usage
### description
### argument
### options
## command detail 2...

note that aztec start is a wonky example

it lives on a command detail level, but maybe we should consider refactoring it to something like

aztec node
aztec sequencer

instead of the current `aztec start --node` `aztec start --sequencer`

This so we can extend the current model across aztec start

This PR adds scripts to generate this and gives an example of the an autogenerated file.

critesjosh commented on docs/scripts/CODE_REVIEW.md:
Are you planning to implement thses suggested changes?

---

Some additional suggestions:

Critical Issues (Must Fix Before Merge)

  🔴 1. Security: Command Injection Risk via Template Loading

  File: docs/scripts/transform_cli_docs.py:240-242Severity: CRITICAL

  The script uses exec_module() to load arbitrary Python files:
  spec.loader.exec_module(config_module)  # Executes arbitrary code

  Risk: If an attacker controls the template file path, they can execute
  arbitrary code.

  Fix: Add path validation to restrict template files to the scripts
  directory.

  ---
  🔴 2. Security: Insecure Temporary File Handling

  Files:
  - docs/scripts/update_cli_docs.sh:21,27,34
  - docs/scripts/update_cli_docs_unified.sh:60,66,73

  Severity: CRITICAL

  Predictable temp file names vulnerable to symlink attacks:
  TEMP_JSON="/tmp/aztec_cli_docs_$(date +%s).json"

  Fix: Use mktemp for secure temp file creation with cleanup traps.

  ---
  🔴 3. Silent Failures: || true Masks Critical Errors

  File: docs/scripts/update_cli_docs.sh (multiple locations)

  Severity: CRITICAL

  The pattern command || true allows failures to pass silently, making
  debugging impossible.

  Fix: Remove || true or replace with proper error handling that logs
  failures.

  ---
  High Priority Issues (Fix Before Merge)

  ⚠️ 4. Missing File I/O Error Handling

  Files: generate_cli_docs.py, transform_cli_docs.py

  Issue: File operations lack try-except blocks. Missing files or permission
   errors will cause unhandled exceptions.

  Fix: Wrap all file I/O operations in try-except blocks with helpful error
  messages.

  ---
  ⚠️ 5. Potential Command Injection via CLI Arguments

  File: docs/scripts/generate_cli_docs.py:198

  Issue: User-controlled command names passed directly to subprocess without
   validation.

  Fix: Add regex validation to ensure command names contain only safe
  characters: ^[a-zA-Z0-9_-]+$

  ---
  ⚠️ 6. Inadequate Shell Script Error Handling

  Files: All shell scripts

  Issue: Scripts use set -e but miss other important flags.

  Fix: Use comprehensive error handling:
  set -euo pipefail  # Exit on error, undefined variable, pipe failure

  ---
  ⚠️ 7. Inefficient Date Command

  File: docs/scripts/generate_cli_docs.py:257

  Issue: Spawns subprocess for date when Python has built-in datetime.

  Fix:
  from datetime import datetime
  "scanned_at": datetime.now().isoformat()

  ---
  ⚠️ 8. Fragile JSON Parsing

  File: docs/scripts/update_cli_docs.sh:85

  Issue: Regex-based JSON parsing will break with formatting changes.

  Fix: Already improved in unified script with jq fallback - use that
  approach consistently.

  ---
  ⚠️ 9. YAML Unavailability Should Be Fatal

  File: docs/scripts/transform_cli_docs.py

  Issue: Returns silently when YAML library is unavailable instead of
  exiting with error code.

  Fix: Exit with proper error code when dependencies are missing.

  ---
  Medium Priority Issues (Should Fix)

  🟡 10. Massive Code Duplication

  Files: update_cli_docs.sh vs update_cli_wallet_docs.sh,
  generate_aztec_docs.sh vs generate_aztec_wallet_docs.sh

  Issue: Near-identical scripts violate DRY principle.

  Recommendation: Use the unified scripts already in the PR and deprecate
  the duplicates.

  ---
  🟡 11. Missing Input Validation

  File: docs/scripts/transform_cli_docs.py

  Issue: No validation of input data structure before processing.

  Fix: Add checks to ensure required keys exist in the JSON data.

  ---
  🟡 12. No Temp File Cleanup

  Files: Shell scripts creating temp files

  Issue: Temp files not cleaned up on script failure.

  Fix: Add trap handlers:
  trap 'rm -f "$TEMP_JSON" "$TEMP_MD"' EXIT

  ---
  🟡 13. Subprocess Error Handling Too Broad

  File: docs/scripts/generate_cli_docs.py

  Issue: Catching all exceptions makes it hard to distinguish command not
  found from other errors.

  Fix: Catch FileNotFoundError separately to provide better error messages.

sklppy88 replied:
Yep, have implemented the ones I think relevant !

critesjosh reviewed (commented):
maybe put all of these scripts into one folder called `cli_generation` or something similar.

Where / when do you think we should run these scripts to autogen the docs? At some point in CI I expect?

sklppy88 commented:
> maybe put all of these scripts into one folder called `cli_generation` or something similar.
> 
> Where / when do you think we should run these scripts to autogen the docs? At some point in CI I expect?

Done, as discussed during the last call, I think we should do these manually before release !

critesjosh reviewed (approved):
looks good! The devnet version of the docs has been merged since you started this. can you run it on the devnet.4 version and add the output pages?

and then just some minor nits from Claude around references to other files using incorrect naming:

  8. Incorrect Script Names in Examples (scan_cli.py:3-7)
    - Docstrings reference generate_cli_docs.py but actual file is scan_cli.py
    - Copy-paste examples will fail
    - Fix: Update all docstrings to use correct filenames
  9. Shell Script Header Comments Wrong (generate_cli_docs.sh:2, update_cli_docs.sh:3)
    - Usage examples reference _unified.sh suffix that doesn't exist
    - Fix: Correct header comments to match actual filenames

sklppy88 committed:
feat(docs): autogenerating cli references spike

This defines both the template, and the method.

The template is simple, and uses one page.

At the entrypoint we have `aztec`

#command (aztec)
## table of contents
## command detail 1
### usage
### description
### argument
### options
## command detail 2...

note that aztec start is a wonky example

it lives on a command detail level, but maybe we should consider refactoring it to something like

aztec node
aztec sequencer

instead of the current `aztec start --node` `aztec start --sequencer`

This so we can extend the current model across aztec start

This PR adds scripts to generate this and gives an example of the an autogenerated file.

Co-authored-by: Esau <esau@aztecprotocol.com>
